test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,410 @@
+
+#include <fcntl.h>
+#include <poll.h>
+
+#include <chrono>
+#include <memory>
+#include <random>
+#include <string>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include "absl/strings/str_cat.h"
+
+#include <grpc/event_engine/event_engine.h>
+
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/event_engine/channel_args_endpoint_config.h"
+#include "src/core/lib/event_engine/posix_engine/event_poller.h"
+#include "src/core/lib/event_engine/posix_engine/event_poller_posix_default.h"
+#include "src/core/lib/event_engine/posix_engine/posix_engine.h"
+#include "src/core/lib/event_engine/posix_engine/posix_engine_closure.h"
+#include "src/core/lib/event_engine/posix_engine/tcp_socket_utils.h"
+#include "src/core/lib/gprpp/dual_ref_counted.h"
+#include "src/core/lib/gprpp/global_config.h"
+#include "src/core/lib/iomgr/port.h"
+#include "test/core/event_engine/test_suite/event_engine_test_utils.h"
+#include "test/core/event_engine/test_suite/oracle_event_engine_posix.h"
+#include "test/core/util/port.h"
+
+GPR_GLOBAL_CONFIG_DECLARE_STRING(grpc_poll_strategy);
+
+namespace grpc_event_engine {
+namespace posix_engine {
+
+using ::grpc_event_engine::experimental::ChannelArgsEndpointConfig;
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::Poller;
+using ::grpc_event_engine::experimental::PosixEventEngine;
+using ::grpc_event_engine::experimental::PosixOracleEventEngine;
+using ::grpc_event_engine::experimental::Promise;
+using ::grpc_event_engine::experimental::URIToResolvedAddress;
+using Endpoint = ::grpc_event_engine::experimental::EventEngine::Endpoint;
+using Listener = ::grpc_event_engine::experimental::EventEngine::Listener;
+using namespace std::chrono_literals;
+
+namespace {
+
+constexpr int kMinMessageSize = 1024;
+constexpr int kMaxMessageSize = 8192;
+constexpr int kNumConnections = 10;
+constexpr int kNumExchangedMessages = 100;
+std::atomic<int> g_num_active_connections{0};
+
+// Returns a random message with bounded length.
+std::string GetNextSendMessage() {
+  static const char alphanum[] =
+      "0123456789"
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+      "abcdefghijklmnopqrstuvwxyz";
+  static std::random_device rd;
+  static std::seed_seq seed{rd()};
+  static std::mt19937 gen(seed);
+  static std::uniform_real_distribution<> dis(kMinMessageSize, kMaxMessageSize);
+  static grpc_core::Mutex g_mu;
+  std::string tmp_s;
+  int len;
+  {
+    grpc_core::MutexLock lock(&g_mu);
+    len = dis(gen);
+  }
+  tmp_s.reserve(len);
+  for (int i = 0; i < len; ++i) {
+    tmp_s += alphanum[rand() % (sizeof(alphanum) - 1)];
+  }
+  return tmp_s;
+}
+
+std::shared_ptr<EventEngine> GetPosixEE() {
+  static std::shared_ptr<EventEngine> posix_ee =
+      std::make_shared<PosixEventEngine>();
+  return posix_ee;
+}
+
+EventEngine* GetOracleEE() {
+  static EventEngine* oracle_ee = new PosixOracleEventEngine();
+  return oracle_ee;
+}
+
+class TestScheduler : public Scheduler {
```

Comment:
This is copy-pasted from another test. Please deduplicate.

Reply:
I created posix_engine_test_utils{.h,.cc} files and moved it there.